### ISSUE TITLE
Unable to run the services using docker-compose

### ISSUE THREAD

**Issue by DazWilkin** [google/keytransparency#1159]
Regrettably (for me), I'm not very familiar with docker-compose and am struggling to run the services using it. Aside from the `${TRAVIS_COMMIT}` issue (#1156}, there appears to be a race-condition with the database and I'm getting errors when running the sequencer.

I've tried paring the compose file back to (a) just running the db;  and (b) then adding in e.g. trillian; but once I (c) add the sequencer, the following issues arise:

- database race-condition
- failed to create etcd client
- address [:8080] already in use
- complaints of handshake errors on multiple localhost:578XX ports

```
docker-compose logs sequencer
Attaching to keytransparency_sequencer_1
sequencer_1   | F0108 22:30:26.169091       1 main.go:82] db.Ping(): dial tcp 172.22.0.2:3306: connect: connection refused
sequencer_1   | F0108 22:30:28.052764       1 main.go:82] db.Ping(): dial tcp 172.22.0.2:3306: connect: connection refused
sequencer_1   | F0108 22:30:29.397395       1 main.go:82] db.Ping(): dial tcp 172.22.0.2:3306: connect: connection refused
sequencer_1   | F0108 22:30:31.812686       1 main.go:82] db.Ping(): dial tcp 172.22.0.2:3306: connect: connection refused
sequencer_1   | F0108 22:30:33.332665       1 main.go:82] db.Ping(): dial tcp 172.22.0.2:3306: connect: connection refused
sequencer_1   | I0108 22:30:35.740407       1 main.go:190] Signer starting
sequencer_1   | I0108 22:30:35.741232       1 server.go:45] Hosting metrics on :8081
sequencer_1   | I0108 22:30:35.741392       1 server.go:57] Listening on [::]:8080
sequencer_1   | E0108 22:30:35.741478       1 server.go:59] ListenAndServeTLS: listen tcp 0.0.0.0:8080: bind: address already in use
sequencer_1   | F0108 22:30:40.741629       1 main.go:199] Failed to create etcd client: dial tcp: missing address
sequencer_1   | I0108 22:30:44.577583       1 main.go:190] Signer starting
sequencer_1   | I0108 22:30:44.578224       1 server.go:45] Hosting metrics on :8081
sequencer_1   | I0108 22:30:44.578286       1 server.go:57] Listening on [::]:8080
sequencer_1   | E0108 22:30:44.578337       1 server.go:59] ListenAndServeTLS: listen tcp 0.0.0.0:8080: bind: address already in use
sequencer_1   | F0108 22:30:49.578537       1 main.go:199] Failed to create etcd client: dial tcp: missing address
sequencer_1   | I0108 22:30:56.592233       1 main.go:190] Signer starting
sequencer_1   | I0108 22:30:56.593104       1 server.go:45] Hosting metrics on :8081
sequencer_1   | I0108 22:30:56.593206       1 server.go:57] Listening on [::]:8080
sequencer_1   | E0108 22:30:56.593243       1 server.go:59] ListenAndServeTLS: listen tcp 0.0.0.0:8080: bind: address already in use
sequencer_1   | I0108 22:31:18.472551       1 main.go:190] Signer starting
sequencer_1   | I0108 22:31:18.473016       1 server.go:45] Hosting metrics on :8081
sequencer_1   | I0108 22:31:18.473104       1 server.go:57] Listening on [::]:8080
sequencer_1   | E0108 22:31:18.473152       1 server.go:59] ListenAndServeTLS: listen tcp 0.0.0.0:8080: bind: address already in use
sequencer_1   | F0108 22:31:23.473405       1 main.go:199] Failed to create etcd client: dial tcp: missing address
sequencer_1   | I0108 22:31:24.423724       1 main.go:190] Signer starting
sequencer_1   | I0108 22:31:24.425749       1 server.go:57] Listening on [::]:8080
sequencer_1   | I0108 22:31:24.425737       1 server.go:45] Hosting metrics on :8081
sequencer_1   | 2019/01/08 22:31:24 http: TLS handshake error from 127.0.0.1:57856: remote error: tls: bad certificate
sequencer_1   | 2019/01/08 22:31:25 http: TLS handshake error from 127.0.0.1:57858: tls: first record does not look like a TLS handshake
sequencer_1   | 2019/01/08 22:31:25 http: TLS handshake error from 127.0.0.1:57860: remote error: tls: bad certificate
sequencer_1   | 2019/01/08 22:31:26 http: TLS handshake error from 127.0.0.1:57862: tls: first record does not look like a TLS handshake
sequencer_1   | 2019/01/08 22:31:27 http: TLS handshake error from 127.0.0.1:57864: remote error: tls: bad certificate
sequencer_1   | F0108 22:31:29.426042       1 main.go:199] Failed to create etcd client: dial tcp: missing address
sequencer_1   | I0108 22:31:30.367661       1 main.go:190] Signer starting
sequencer_1   | I0108 22:31:30.368858       1 server.go:57] Listening on [::]:8080
sequencer_1   | I0108 22:31:30.368865       1 server.go:45] Hosting metrics on :8081
sequencer_1   | 2019/01/08 22:31:30 http: TLS handshake error from 127.0.0.1:57918: remote error: tls: bad certificate
sequencer_1   | 2019/01/08 22:31:31 http: TLS handshake error from 127.0.0.1:57920: tls: first record does not look like a TLS handshake
sequencer_1   | 2019/01/08 22:31:31 http: TLS handshake error from 127.0.0.1:57922: remote error: tls: bad certificate
```

**Comment by DazWilkin** [comment]
The issue appears (!?) to be with the overloading of the listening (addr) port [8080] for both the gRPC and HTTP services. I don't understand the code sufficiently to understand what's going wrong but, the difference between how the sequencer and the server (simpler) do this, appears to include the difference.

I had thought that the `bind: address already in use` error was some conflict between the docker-compose containers but, no matter what combination of ports I use, the issue arise directly from within the binary; the issue arises when running the code outside of the container too.

**Comment by gdbelvin** [comment]
Yes, this multiplexing error was kind subtle.  I just fixed it in #1164

**Comment by DazWilkin** [comment]
I'll try this again when I'm back from vacation and close the issue once I have it working.

Thanks for addressing it.
